syft/execution/protocol.py

```diff
@@ -1,479 +1,560 @@
+    #         # We need to deepcopy here otherwise the real layer is modified when the Protocol is being built
+    #         copied_layer = copy.deepcopy(value)
+    #         for copied_param, param in zip(copied_layer.named_parameters(), value.parameters()):
+    #             (copied_name, _) = copied_param
+    #             copied_layer._parameters[copied_name] = self.role.placeholders[param.id]
+
+    #         return copied_layer
+
+    def __call__(self, *args):
         """
-        Run the protocol by executing the plans sequentially
+        Calls a protocol execution with some arguments.
+
+        When possible, run the original function to improve efficiency. When
+        it's not, for example if you fetched the protocol from a remote worker,
+        then run it from the tape of actions:
+        - Instantiate input placeholders
+        - for each recorded action, run the action on the placeholders
+          and use the result(s) to instantiate to appropriate placeholder.
+        - Return the instantiation of all the output placeholders.
+        """
+        results_per_role = {}
+        for role_id, role in self.roles.items():
+            args_for_role = [arg for arg in args if arg.owner == role_id]
+            results_per_role[role_id] = role.execute(args_for_role)
+
+        return results_per_role
 
-        The input args_ provided are sent to the first plan location. This first plan is
-        run and its output is moved to the second plan location, and so on. The final
-        result is returned after all plans have run, and it is composed of pointers to
-        the last plan location.
+    def run(self, args_: Tuple, result_ids: List[Union[str, int]]):
+        """Controls local or remote protocol execution.
+        If the protocol doesn't have the protocol built, first build it using the original function.
 
-        Raises:
-            RuntimeError: If the protocol has a location attribute and it is not
-                the local worker
+        Args:
+            args_: Arguments used to run protocol.
+            result_ids: List of ids where the results will be stored.
         """
-        self._assert_is_resolved()
-
-        # This is an alternative to having PointerProtocol, when we send the protocol.
-        if self.location is not None:
-            location = Protocol.find_args_location(args)
-            if location != self.location:
-                raise RuntimeError(
-                    f"This protocol has been sent to {self.location.id}, but you provided "
-                    f"local arguments or pointers to {location.id}."
-                )
+        # TODO: can we reuse result_ids?
+        return self.__call__(*args_)
 
-            print("send remote run request to", self.location.id)
-            response = self.request_remote_run(location, args, kwargs)
-            return response
+    # def send(self, *locations: AbstractWorker, force=False):
```

Comment:
Let's vaporize anything that's commented out in this PR. We can always add this stuff back later, since it came from `Plan` anyway.